### PR TITLE
[Exporter.StackDriver] Add net8 and net6 TFM support and update Google Cloud libraries (#1414)

### LINKED ISSUE #1414: Target .NETStandard 2.1 or runtimes explicitly to get rid of dependency on a native Grpc.Core library that may conflict with gRPC-dotnet
# Issue with OpenTelemetry.Exporter.Stackdriver

Please, also target .NETStandard 2.1 or even the latest .NET runtimes explicitly to eliminate dependency on the native `Grpc.Core` library that may add conflicts to applications that also target a native .NET-based gRPC implementation. Please see [gRPC C# docs](https://grpc.io/docs/languages/csharp/)

Version:

1.0.0-beta.4

Runtime version net6.0+

**Is this a feature request or a bug?**

* [ ] Feature

**What is the expected behavior?**
For modern runtimes, the system should resolve transitive dependencies of Google SDK libraries, so it uses [gRPC-dotnet](https://github.com/grpc/grpc-dotnet).

What do you expect to see?

For applications that target .NETStandard 2.1 and above, the build process should use a "native" .NET-based gRPC implementation.

**What is the actual behavior?**

On the one hand, when an app adds `OpenTelemetry.Exporter.Stackdriver` as a dependency
```
    <PackageReference Include="OpenTelemetry.Exporter.Stackdriver" Version="1.0.0-beta.4" />
```
The library brings a native `Grpc.Core` library that relies on p/Invoke as a transitive dependency.

```
dotnet list package  --include-transitive|grep Grpc
   > Google.Api.Gax.Grpc                                        3.7.0
   > Google.Api.Gax.Grpc.GrpcCore                               3.7.0
   > Grpc.Auth                                                  2.41.0
   > Grpc.Core                                                  2.41.0
   > Grpc.Core.Api                                              2.41.0
```

On the other hand, if the other application targets .NET6.0+ and adds  Google SDK library explicitly
```
<PackageReference Include="Google.Cloud.Diagnostics.Common" Version="5.1.0" />
```
A more modern and recommended gRPC-dotnet library is resolved
```
dotnet list package  --include-transitive|grep Grpc
   > Google.Api.Gax.Grpc                                        4.3.1
   > Grpc.Auth                                                  2.51.0
   > Grpc.Core.Api                                              2.51.0
   > Grpc.Net.Client                                            2.51.0
   > Grpc.Net.Common                                            2.51.0
```

**The issue**:
If the app targets one of the Google Libraries `Google.Cloud.*` and `OpenTelemetry.Exporter.Stackdriver`, it brings some instability to the app and may result in runtime binding exceptions (missing methods, etc.).

### PR DESCRIPTION
**Pull request by StasPerekrestov**
Fixes #1414 

## Changes

Adds support for netstandard2.1 and updates dependency on Google .NET SDK libraries, so applications that use .NET 6+ resolve natively implemented gRPC libraries.

Google Release notes
[Google.Cloud.Monitoring.V3 version 3.0.0](https://github.com/googleapis/google-cloud-dotnet/pull/8682)

[Google.Cloud.Trace.V2 version 3.0.0](https://github.com/googleapis/google-cloud-dotnet/pull/8618)

[GAX v4 changes](https://cloud.google.com/dotnet/docs/reference/help/breaking-gax4)

### CODE REVIEW

**Review by cijothomas — APPROVED** [review]

**Comment by Kielek** [comment]
@SergeyKanzhelev please check

File: src/OpenTelemetry.Exporter.Stackdriver/CHANGELOG.md [hunk 1]
```diff
@@ -4,6 +4,8 @@
 
 * Update OpenTelemetry SDK version to `1.7.0`.
   ([#1486](https://github.com/open-telemetry/opentelemetry-dotnet-contrib/pull/1486))
+* Add netstandard2.1 support and update Google Cloud libraries.
```
Comment (Kielek) [inline src/OpenTelemetry.Exporter.Stackdriver/CHANGELOG.md:7]: Could you please put exact versions here?

File: src/OpenTelemetry.Exporter.Stackdriver/OpenTelemetry.Exporter.Stackdriver.csproj [hunk 2]
```diff
@@ -1,16 +1,16 @@
 <Project Sdk="Microsoft.NET.Sdk">
   <PropertyGroup>
     <!-- OmniSharp/VS Code requires TargetFrameworks to be in descending order for IntelliSense and analysis. -->
-    <TargetFrameworks>netstandard2.0;net462</TargetFrameworks>
+    <TargetFrameworks>netstandard2.0;net462;netstandard2.1</TargetFrameworks>
```
Comment (Kielek) [inline src/OpenTelemetry.Exporter.Stackdriver/OpenTelemetry.Exporter.Stackdriver.csproj:4]: Please reorder to
```suggestion
    <TargetFrameworks>netstandard2.1;netstandard2.0;net462</TargetFrameworks>
```

or consider. I do not see any good reason to keep both netstanard2.0 and netstandard2.1.
Following set should cover all currently supported .NET versions.
```suggestion
    <TargetFrameworks>net6.0;net462</TargetFrameworks>
```

File: src/OpenTelemetry.Exporter.Stackdriver/CHANGELOG.md (hunk 1 above)
Comment (StasPerekrestov) [reply src/OpenTelemetry.Exporter.Stackdriver/CHANGELOG.md:7]: Should I specify the exact versions of two Google SDK libs?

File: src/OpenTelemetry.Exporter.Stackdriver/OpenTelemetry.Exporter.Stackdriver.csproj (hunk 2 above)
Comment (StasPerekrestov) [reply src/OpenTelemetry.Exporter.Stackdriver/OpenTelemetry.Exporter.Stackdriver.csproj:4]: What about net8?

Comment (Kielek) [reply src/OpenTelemetry.Exporter.Stackdriver/OpenTelemetry.Exporter.Stackdriver.csproj:4]: As I understand, TFMs, you can reference net6 in net7 and net8. Probably also net9.

File: src/OpenTelemetry.Exporter.Stackdriver/CHANGELOG.md (hunk 1 above)
Comment (Kielek) [reply src/OpenTelemetry.Exporter.Stackdriver/CHANGELOG.md:7]: Yes please. It can be helful for endusers.

**Comment by StasPerekrestov** [comment]
@Kielek, could you please suggest if I should include code analysis-related fixes in the PR? There are some issues that break the build.

```
Error: D:\a\opentelemetry-dotnet-contrib\opentelemetry-dotnet-contrib\src\OpenTelemetry.Exporter.Stackdriver\StackdriverTraceExporter.cs(44,44): error CS8602: Dereference of a possibly null reference. [D:\a\opentelemetry-dotnet-contrib\opentelemetry-dotnet-contrib\src\OpenTelemetry.Exporter.Stackdriver\OpenTelemetry.Exporter.Stackdriver.csproj::TargetFramework=net8.0]
Error: D:\a\opentelemetry-dotnet-contrib\opentelemetry-dotnet-contrib\src\OpenTelemetry.Exporter.Stackdriver\Implementation\ActivityExtensions.cs(49,13): error CS8073: The result of the expression is always 'true' since a value of type 'ActivitySpanId' is never equal to 'null' of type 'ActivitySpanId?' [D:\a\opentelemetry-dotnet-contrib\opentelemetry-dotnet-contrib\src\OpenTelemetry.Exporter.Stackdriver\OpenTelemetry.Exporter.Stackdriver.csproj::TargetFramework=net8.0]
Error: D:\a\opentelemetry-dotnet-contrib\opentelemetry-dotnet-contrib\src\OpenTelemetry.Exporter.Stackdriver\Utils\CommonUtils.cs(22,9): error CA1510: Use 'ArgumentNullException.ThrowIfNull' instead of explicitly throwing a new exception instance (https://learn.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca1510) [D:\a\opentelemetry-dotnet-contrib\opentelemetry-dotnet-contrib\src\OpenTelemetry.Exporter.Stackdriver\OpenTelemetry.Exporter.Stackdriver.csproj::TargetFramework=net8.0]
```

File: src/OpenTelemetry.Exporter.Stackdriver/OpenTelemetry.Exporter.Stackdriver.csproj [hunk 3]
```diff
@@ -1,16 +1,16 @@
 <Project Sdk="Microsoft.NET.Sdk">
   <PropertyGroup>
     <!-- OmniSharp/VS Code requires TargetFrameworks to be in descending order for IntelliSense and analysis. -->
-    <TargetFrameworks>netstandard2.0;net462</TargetFrameworks>
+    <TargetFrameworks>net8.0;net7.0;net6.0;net462</TargetFrameworks>
```
Comment (Kielek) [inline src/OpenTelemetry.Exporter.Stackdriver/OpenTelemetry.Exporter.Stackdriver.csproj:4]: I am not sure if you really need here n3t7.0 and net8.0.

You can reference .NET6 from .NET8.

**Comment by Kielek** [comment]
> @Kielek, could you please suggest if I should include code analysis-related fixes in the PR? There are some issues that break the build.
> 
> ```
> Error: D:\a\opentelemetry-dotnet-contrib\opentelemetry-dotnet-contrib\src\OpenTelemetry.Exporter.Stackdriver\StackdriverTraceExporter.cs(44,44): error CS8602: Dereference of a possibly null reference. [D:\a\opentelemetry-dotnet-contrib\opentelemetry-dotnet-contrib\src\OpenTelemetry.Exporter.Stackdriver\OpenTelemetry.Exporter.Stackdriver.csproj::TargetFramework=net8.0]
> Error: D:\a\opentelemetry-dotnet-contrib\opentelemetry-dotnet-contrib\src\OpenTelemetry.Exporter.Stackdriver\Implementation\ActivityExtensions.cs(49,13): error CS8073: The result of the expression is always 'true' since a value of type 'ActivitySpanId' is never equal to 'null' of type 'ActivitySpanId?' [D:\a\opentelemetry-dotnet-contrib\opentelemetry-dotnet-contrib\src\OpenTelemetry.Exporter.Stackdriver\OpenTelemetry.Exporter.Stackdriver.csproj::TargetFramework=net8.0]
> Error: D:\a\opentelemetry-dotnet-contrib\opentelemetry-dotnet-contrib\src\OpenTelemetry.Exporter.Stackdriver\Utils\CommonUtils.cs(22,9): error CA1510: Use 'ArgumentNullException.ThrowIfNull' instead of explicitly throwing a new exception instance (https://learn.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca1510) [D:\a\opentelemetry-dotnet-contrib\opentelemetry-dotnet-contrib\src\OpenTelemetry.Exporter.Stackdriver\OpenTelemetry.Exporter.Stackdriver.csproj::TargetFramework=net8.0]
> ```

We cannot merge without fixing these errors. I see two possibilities for solving
1. Include it here, only if PR still will be manageable and will not explode with additional changes.
2. Create separate PR which can be merged first, then we can rebase/merge main here.

Based on numbers of issues, I would start with option 1.

BTW for the null checking, and throwing exception, there is `Guard` class, which support also .NET Framework 4.6.2.

File: src/OpenTelemetry.Exporter.Stackdriver/CHANGELOG.md [hunk 4]
```diff
@@ -4,6 +4,13 @@
 
 * Update OpenTelemetry SDK version to `1.7.0`.
   ([#1486](https://github.com/open-telemetry/opentelemetry-dotnet-contrib/pull/1486))
+* Add support of a native "gRPC for .NET" for apps targeting .NET Core 3.0 or later.
+  ([#1414](https://github.com/open-telemetry/opentelemetry-dotnet-contrib/issues/1414))
+  1. Add support net8.0, net7.0, net6.0 as target frameworks.
```
Comment (Kielek) [inline src/OpenTelemetry.Exporter.Stackdriver/CHANGELOG.md:9]:
```suggestion
* Add support of a native "gRPC for .NET" for apps targeting .NET 6.0 or later.
  ([#1414](https://github.com/open-telemetry/opentelemetry-dotnet-contrib/issues/1414))
  1. Add support net8.0, net6.0 as target frameworks.
```

**Comment by Kielek** [comment]
@SergeyKanzhelev, please check.

**Comment by StasPerekrestov** [comment]
Hello. Please let me know if something needs to be changed.
I'll gladly address all the issues.
Thank you.

**Comment by StasPerekrestov** [comment]
Hello @cijothomas and @Kielek,

I wanted to kindly remind you that the pull request has been marked as stale and is at risk of being closed.

If you have any questions, please don't hesitate to ask.

Thank you.

**Review by SergeyKanzhelev — APPROVED** [review]

**Pull request merged**